Documentation/Evolution/StronglyTypedCaptures.md

```diff
@@ -0,0 +1,647 @@
+//         Optionally {
+//             ".."
+//             OneOrMore(CharacterClass.hexDigit).capture()
+//         }
+//     }
+//     // `.Captures == (Substring, Substring?)`
+```
+
+#### Nested capturing group
+
+When capturing group is nested within another capturing group, they count as two
+distinct captures in the order their left parenthesis first appears in the
+regular expression literal. This is consistent with PCRE and allows us to use
+backreferences (e.g. `\2`) with linear indices.
+
+```swift
+let graphemeBreakPropertyData = /(([0-9A-F]+)(\.\.([0-9A-F]+)))\s*;\s(\w+).*/
+// Positions in capture tuple: 0 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    4 ^~~~~
+//                                         2 ^~~~~~~~~~~~~~~~~
+//                              1 ^~~~~~~~~~~   3 ^~~~~~~~~~~
+// => `Regex<(Substring, Substring, Substring, Substring, Substring)>`
+
+// Result builder DSL equivalent:
+//     let graphemeBreakPropertyData = Regex {
+//         Regex {
+//             OneOrMore(CharacterClass.hexDigit).capture() // (1)
+//             Regex {
+//                 ".."
+//                 OneOrMore(CharacterClass.hexDigit).capture() // (3)
+//             }.capture() // (2)
+//         }.capture() // (0)
+//         Repeat(CharacterClass.whitespace)
+//         ";"
+//         CharacterClass.whitespace
+//         OneOrMore(CharacterClass.word).capture() // (4)
+//         Repeat(CharacterClass.any)
+//     }
+//     // `.Captures == Regex(Substring, Substring, Substring, Substring, Substring)`
+
+let input = "007F..009F    ; Control # Cc  [33] <control-007F>..<control-009F>"
+// Match result for `input`:
+// ("007F..009F", "007F", "..009F", "009F", "Control")
+```
+
+
+#### Quantification
+
+A quantifier's capture type depends on its underlying pattern. The laziness of
+a quantifier, as denoted by a `?` suffix, does not affect the capture type. 
+
+```
+pattern ::= pattern '*' '?'?
+pattern ::= pattern '+' '?'?
+pattern ::= pattern '?' '?'?
+pattern ::= pattern '{' number (',' spaces? number?)? '}' '?'?
+```
+
+| Syntax               | Description           | Capture type                                                  |
+| -------------------- | --------------------- | ------------------------------------------------------------- |
+| `*` / `*?`           | 0 or more             | `Array` of the sub-pattern capture type                       |
+| `+` / `+?`           | 1 or more             | `Array` of the sub-pattern capture type                       |
+| `?` / `??`           | 0 or 1                | `Optional` of the sub-pattern capture type                    |
+| `{n}` / `{n}?`       | Exactly _n_           | `Array` of the sub-pattern capture type                       |
+| `{n, m}` / `{n, m}?` | Between _n_ and _m_   | `Array` of the sub-pattern capture type                       |
+| `{n,}` / `{n,}?`     | _n_ or more           | `Array` of the sub-pattern capture type                       |
+
```

Comment:
Hmm here I just wanted to give a clear association between the syntax and the corresponding typing.

Reply:
I think Michael's saying the addition of the lazy quantifiers is really emphasizing the regex syntax. It might be just as clear if in the table we omit the lazy quantifier variants.

Reply:
Removed syntax details as suggested.